src/couch_views/src/couch_views_indexer.erl

```diff
@@ -0,0 +1,262 @@
+            doc_acc := DocAcc,
+            last_seq := LastSeq
+        } = State2,
+
+        {Mrst1, MappedDocs} = map_docs(Mrst0, DocAcc),
+        write_docs(TxDb, Mrst1, MappedDocs, State2),
+
+        case Count < Limit of
+            true ->
+                report_progress(State2, finished),
+                {Mrst1, finished};
+            false ->
+                report_progress(State2, update),
+                {Mrst1, State2#{
+                    tx_db := undefined,
+                    count := 0,
+                    doc_acc := [],
+                    view_seq := LastSeq
+                }}
+        end
+    end),
+
+    case State3 of
+        finished ->
+            couch_query_servers:stop_doc_map(Mrst2#mrst.qserver);
+        _ ->
+            update(Db, Mrst2, State3)
+    end.
+
+
+fold_changes(State) ->
+    #{
+        view_seq := SinceSeq,
+        limit := Limit,
+        tx_db := TxDb
+    } = State,
+
+    Fun = fun process_changes/2,
+    fabric2_db:fold_changes(TxDb, SinceSeq, Fun, State, [{limit, Limit}]).
+
+
+process_changes(Change, Acc) ->
+    #{
+        doc_acc := DocAcc,
+        count := Count,
+        tx_db := TxDb,
+        design_opts := DesignOpts
+    } = Acc,
+
+    #{
+        id := Id,
+        sequence := LastSeq,
+        deleted := Deleted
+    } = Change,
+
+    IncludeDesign = lists:keymember(<<"include_design">>, 1, DesignOpts),
+
+    Acc1 = case {Id, IncludeDesign} of
+        {<<"_design/", _/binary>>, false} ->
+            maps:merge(Acc, #{
+                count => Count + 1,
+                last_seq => LastSeq
+            });
+        _ ->
+            % Making a note here that we should make fetching all the docs
+            % a parallel fdb operation
+            {ok, Doc} = case Deleted of
+                true -> {ok, []};
+                false -> fabric2_db:open_doc(TxDb, Id)
+            end,
+
+            Change1 = maps:put(doc, Doc, Change),
+            Acc#{
+                doc_acc := DocAcc ++ [Change1],
+                count := Count + 1,
+                last_seq := LastSeq
+            }
+    end,
+    {ok, Acc1}.
+
+
+map_docs(Mrst, Docs) ->
+    % Run all the non deleted docs through the view engine and
+    Mrst1 = start_query_server(Mrst),
+    QServer = Mrst1#mrst.qserver,
+    MapFun = fun
+        (#{deleted := true} = Change) ->
+            maps:put(results, [], Change);
+        (Change) ->
+            #{doc := Doc} = Change,
+            couch_stats:increment_counter([couchdb, mrview, map_doc]),
+            {ok, RawResults} = couch_query_servers:map_doc_raw(QServer, Doc),
+            JsonResults = couch_query_servers:raw_to_ejson(RawResults),
+            ListResults = lists:map(fun(ViewResults) ->
+                [list_to_tuple(Res) || Res <- ViewResults]
+            end, JsonResults),
+            maps:put(results, ListResults, Change)
+    end,
+    {Mrst1, lists:map(MapFun, Docs)}.
+
+
+write_docs(TxDb, Mrst, Docs, State) ->
+    #mrst{
```

Comment:
(optional) could unpack #mrst and State in 2 lines respectively instead of 8 lines, so save on some vertical space

Reply:
I've actually quite enjoyed using a few extra lines. I think it comes from me writing to much javascript